### PR TITLE
hnsw: add neon simd support to l2 and dot product

### PR DESCRIPTION
**Pull request by asdine**
### What's being changed:

This adds support for NEON SIMD instructions for the L2 and Dot product calculations.

Overall, the calculation are twice as fast compared to the Go version. 🎉 

Benchmark results:

L2:

```
                                     │  l2_go.txt   │            l2_neon.txt             │
                                     │    sec/op    │   sec/op     vs base               │
_L2_PureGo_VS_Neon/vector_dim=30-8     10.690n ± 4%   6.292n ± 1%  -41.15% (p=0.002 n=6)
_L2_PureGo_VS_Neon/vector_dim=32-8     11.320n ± 2%   5.974n ± 4%  -47.22% (p=0.002 n=6)
_L2_PureGo_VS_Neon/vector_dim=128-8     48.22n ± 8%   17.18n ± 5%  -64.38% (p=0.002 n=6)
_L2_PureGo_VS_Neon/vector_dim=256-8     89.29n ± 2%   39.25n ± 0%  -56.04% (p=0.002 n=6)
_L2_PureGo_VS_Neon/vector_dim=300-8    103.00n ± 9%   46.60n ± 7%  -54.76% (p=0.002 n=6)
_L2_PureGo_VS_Neon/vector_dim=384-8    129.00n ± 6%   59.73n ± 0%  -53.69% (p=0.002 n=6)
_L2_PureGo_VS_Neon/vector_dim=600-8    197.10n ± 0%   99.44n ± 5%  -49.55% (p=0.002 n=6)
_L2_PureGo_VS_Neon/vector_dim=768-8     249.3n ± 0%   139.5n ± 0%  -44.05% (p=0.002 n=6)
_L2_PureGo_VS_Neon/vector_dim=1024-8    329.4n ± 1%   200.1n ± 0%  -39.24% (p=0.002 n=6)
geomean                                 78.41n        38.72n       -50.61%
```

Dot:
```
                      │  dot_go.txt  │            dot_neon.txt            │
                      │    sec/op    │   sec/op     vs base               │
Dot/30_dimensions-8     10.695n ± 6%   6.177n ± 6%  -42.24% (p=0.002 n=6)
Dot/32_dimensions-8     11.270n ± 0%   5.811n ± 3%  -48.44% (p=0.002 n=6)
Dot/128_dimensions-8     48.35n ± 0%   16.48n ± 0%  -65.91% (p=0.002 n=6)
Dot/256_dimensions-8     89.22n ± 0%   38.70n ± 5%  -56.63% (p=0.002 n=6)
Dot/300_dimensions-8    102.85n ± 5%   44.86n ± 1%  -56.38% (p=0.002 n=6)
Dot/384_dimensions-8    129.30n ± 0%   58.47n ± 6%  -54.78% (p=0.002 n=6)
Dot/600_dimensions-8    196.90n ± 0%   98.39n ± 0%  -50.03% (p=0.002 n=6)
Dot/768_dimensions-8     249.7n ± 5%   139.8n ± 5%  -44.02% (p=0.002 n=6)
Dot/1024_dimensions-8    331.1n ± 0%   201.1n ± 8%  -39.25% (p=0.002 n=6)
geomean                  78.45n        38.02n       -51.53%
```

Fixes https://github.com/weaviate/weaviate/issues/3204
### Review checklist

- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [x] All new code is covered by tests where it is reasonable.
- [x] Performance tests have been run or not necessary.

### CODE REVIEW

File: adapters/repos/db/vector/hnsw/distancer/dot_product_arm64_test.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+	}
+	fmt.Printf("total failed: %d\n", countFailed)
+}
+
+func TestCompareDotProductImplementations(t *testing.T) {
+	sizes := []uint{
+		8,
+		16,
+		32,
```
Comment (trengrj) [inline adapters/repos/db/vector/hnsw/distancer/dot_product_arm64_test.go:99]: Would be nice to have some sizes not divisible by 4 similar to `Test_L2_DistanceImplementation`

**Comment by trengrj** [comment]
Some tests on my M1 with the `ann-benchmarks` chaos pipelines. Results look good:

SIFT1m (l2) comparison
sift1m-neon

Glove Angular (cosine / normalized dot product) comparison
glove-1m-neon

**Comment by trengrj** [comment]
I did an additional test using an AWS Graviton 2 [m6g](https://aws.amazon.com/ec2/instance-types/m6g/) instance and didn't get a performance improvement similar to the M1. Do we need to be using Graviton 3 instances for all the neon features? `cpu.ARM64.HasASIMD` is returning true.

glove-1m-graviton

```bash
$ go version
go version go1.20.5 linux/arm64
$ cat query.go 
package main

import (
        "fmt"
        "golang.org/x/sys/cpu"
)

func main() {
        fmt.Printf("HasASIMD: %v\n", cpu.ARM64.HasASIMD)
}

$ go run query.go 
HasASIMD: true

$ cat /proc/cpuinfo
processor       : 0
BogoMIPS        : 243.75
Features        : fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm lrcpc dcpop asimddp ssbs
CPU implementer : 0x41
CPU architecture: 8
CPU variant     : 0x3
CPU part        : 0xd0c
CPU revision    : 1

processor       : 1
BogoMIPS        : 243.75
Features        : fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm lrcpc dcpop asimddp ssbs
CPU implementer : 0x41
CPU architecture: 8
CPU variant     : 0x3
CPU part        : 0xd0c
CPU revision    : 1
...
```

**Comment by asdine** [comment]
Here is the result on AWS Graviton with the sphere-m1 test:

output

**Comment by asdine** [comment]
Result with compression on the sift dataset:


```
python3 run.py -v /home/ubuntu/sift-128-euclidean.hdf5 -d l2-squared -m 16 -c --dim-to-segment-ratio 4 -l run=compressed --labels "run=neon,dataset=sphere,machine_type=graviton3"
```

output

**Comment by asdine** [comment]
I managed to optimize the l2 implementation by a lot for dim > 32 

```
pkg: github.com/weaviate/weaviate/adapters/repos/db/vector/hnsw/distancer
                                     │  purego.txt  │               v1.txt                │               v16.txt               │
                                     │    sec/op    │    sec/op     vs base               │    sec/op     vs base               │
_L2_PureGo_VS_Neon/vector_dim=30-8     10.670n ± 8%    6.366n ± 6%  -40.33% (p=0.002 n=6)   6.657n ±  1%  -37.61% (p=0.002 n=6)
_L2_PureGo_VS_Neon/vector_dim=32-8     11.300n ± 5%    5.975n ± 0%  -47.12% (p=0.002 n=6)   6.377n ±  1%  -43.57% (p=0.002 n=6)
_L2_PureGo_VS_Neon/vector_dim=128-8     48.21n ± 5%    17.22n ± 9%  -64.28% (p=0.002 n=6)   14.06n ±  5%  -70.83% (p=0.002 n=6)
_L2_PureGo_VS_Neon/vector_dim=256-8     89.11n ± 0%    39.80n ± 2%  -55.33% (p=0.002 n=6)   24.29n ±  0%  -72.75% (p=0.002 n=6)
_L2_PureGo_VS_Neon/vector_dim=300-8    103.15n ± 5%    46.87n ± 4%  -54.57% (p=0.002 n=6)   27.78n ±  5%  -73.07% (p=0.002 n=6)
_L2_PureGo_VS_Neon/vector_dim=384-8    129.40n ± 0%    59.79n ± 1%  -53.79% (p=0.002 n=6)   34.46n ±  1%  -73.37% (p=0.002 n=6)
_L2_PureGo_VS_Neon/vector_dim=600-8    199.55n ± 3%   100.28n ± 9%  -49.74% (p=0.002 n=6)   51.71n ±  7%  -74.09% (p=0.002 n=6)
_L2_PureGo_VS_Neon/vector_dim=768-8    249.35n ± 2%   139.50n ± 2%  -44.05% (p=0.002 n=6)   65.10n ±  1%  -73.89% (p=0.002 n=6)
_L2_PureGo_VS_Neon/vector_dim=1024-8   329.45n ± 3%   201.75n ± 5%  -38.76% (p=0.002 n=6)   85.55n ± 12%  -74.03% (p=0.002 n=6)
geomean                                 78.51n         38.95n       -50.39%                 25.13n        -67.99%
```

**Comment by asdine** [comment]
Same improvement for dot with dim > 32:

```
goos: darwin
goarch: arm64
pkg: github.com/weaviate/weaviate/adapters/repos/db/vector/hnsw/distancer
                      │ dot_pure_go.txt │              dot_v1.txt              │             dot_v2.txt              │
                      │     sec/op      │    sec/op      vs base               │    sec/op     vs base               │
Dot/30_dimensions-8       10.660n ±  2%    6.223n ± 13%  -41.62% (p=0.002 n=6)   6.379n ± 39%  -40.15% (p=0.002 n=6)
Dot/32_dimensions-8       11.540n ± 10%    5.908n ±  7%  -48.80% (p=0.002 n=6)   6.229n ±  0%  -46.02% (p=0.002 n=6)
Dot/128_dimensions-8       48.44n ±  0%    16.57n ± 14%  -65.79% (p=0.002 n=6)   12.00n ±  0%  -75.21% (p=0.002 n=6)
Dot/256_dimensions-8       89.81n ± 17%    39.03n ± 15%  -56.55% (p=0.002 n=6)   19.66n ± 11%  -78.12% (p=0.002 n=6)
Dot/300_dimensions-8      103.35n ±  1%    46.01n ±  2%  -55.48% (p=0.002 n=6)   22.09n ±  0%  -78.63% (p=0.002 n=6)
Dot/384_dimensions-8      129.60n ± 11%    58.81n ± 10%  -54.62% (p=0.002 n=6)   27.24n ±  7%  -78.98% (p=0.002 n=6)
Dot/600_dimensions-8      197.10n ±  0%    99.56n ±  3%  -49.49% (p=0.002 n=6)   39.90n ±  1%  -79.76% (p=0.002 n=6)
Dot/768_dimensions-8      250.60n ± 12%   145.30n ± 16%  -42.02% (p=0.002 n=6)   50.11n ± 14%  -80.00% (p=0.002 n=6)
Dot/1024_dimensions-8     330.85n ±  3%   202.40n ±  1%  -38.82% (p=0.002 n=6)   65.46n ±  0%  -80.22% (p=0.002 n=6)
geomean                    78.80n          38.56n        -51.07%                 20.83n        -73.57%
```

**Comment by asdine** [comment]
Result with compression on the sift dataset:

```
python3 run.py -v /home/ubuntu/sift-128-euclidean.hdf5 -d l2-squared -m 16 -c --dim-to-segment-ratio 4 -l run=compressed --labels "run=neon,dataset=sphere,machine_type=graviton3"
```

output

**Comment by asdine** [comment]
The speed for smaller vectors has been slightly improved:

```
goos: darwin
goarch: arm64
pkg: github.com/weaviate/weaviate/adapters/repos/db/vector/hnsw/distancer
                                     │   l2_go.txt   │          l2_v3_inline.txt           │
                                     │    sec/op     │    sec/op     vs base               │
_L2_PureGo_VS_Neon/vector_dim=2-8       1.885n ± 11%   2.079n ±  1%  +10.29% (p=0.039 n=6)
_L2_PureGo_VS_Neon/vector_dim=4-8       2.506n ±  3%   2.245n ±  5%  -10.43% (p=0.002 n=6)
_L2_PureGo_VS_Neon/vector_dim=6-8       3.139n ± 11%   2.686n ±  2%  -14.43% (p=0.002 n=6)
_L2_PureGo_VS_Neon/vector_dim=8-8       3.756n ± 10%   3.026n ±  1%  -19.46% (p=0.002 n=6)
_L2_PureGo_VS_Neon/vector_dim=10-8      4.382n ±  1%   3.851n ±  7%  -12.13% (p=0.002 n=6)
_L2_PureGo_VS_Neon/vector_dim=12-8      5.054n ± 13%   4.241n ±  0%  -16.08% (p=0.002 n=6)
_L2_PureGo_VS_Neon/vector_dim=16-8      6.279n ±  0%   5.179n ±  4%  -17.51% (p=0.002 n=6)
_L2_PureGo_VS_Neon/vector_dim=24-8      8.777n ±  0%   5.783n ±  1%  -34.12% (p=0.002 n=6)
_L2_PureGo_VS_Neon/vector_dim=30-8     10.655n ± 32%   7.355n ±  0%  -30.97% (p=0.002 n=6)
_L2_PureGo_VS_Neon/vector_dim=32-8     11.270n ±  0%   6.402n ±  0%  -43.19% (p=0.002 n=6)
_L2_PureGo_VS_Neon/vector_dim=128-8     48.20n ±  6%   14.11n ±  4%  -70.72% (p=0.002 n=6)
_L2_PureGo_VS_Neon/vector_dim=256-8     89.11n ±  0%   24.36n ±  0%  -72.66% (p=0.002 n=6)
_L2_PureGo_VS_Neon/vector_dim=300-8    103.20n ±  6%   27.91n ±  2%  -72.96% (p=0.002 n=6)
_L2_PureGo_VS_Neon/vector_dim=384-8    129.05n ±  1%   34.63n ±  1%  -73.17% (p=0.002 n=6)
_L2_PureGo_VS_Neon/vector_dim=600-8    197.55n ±  9%   52.05n ±  3%  -73.65% (p=0.002 n=6)
_L2_PureGo_VS_Neon/vector_dim=768-8    250.25n ± 11%   65.54n ±  0%  -73.81% (p=0.002 n=6)
_L2_PureGo_VS_Neon/vector_dim=1024-8   329.70n ±  0%   85.87n ± 14%  -73.96% (p=0.002 n=6)
geomean                                 19.37n         9.905n        -48.87%
```

**Comment by asdine** [comment]
Results without compression with sift:

output

**Comment by asdine** [comment]
Dot: The speed of smaller vectors has been improved:

```
goos: darwin
goarch: arm64
pkg: github.com/weaviate/weaviate/adapters/repos/db/vector/hnsw/distancer
                      │  dot_go.txt  │            dot_neon.txt             │
                      │    sec/op    │    sec/op     vs base               │
Dot/2_dimensions-8       1.880n ± 5%   2.070n ±  1%  +10.14% (p=0.002 n=6)
Dot/4_dimensions-8       2.504n ± 3%   2.126n ±  3%  -15.10% (p=0.002 n=6)
Dot/6_dimensions-8       3.131n ± 3%   2.559n ±  7%  -18.27% (p=0.002 n=6)
Dot/8_dimensions-8       3.763n ± 3%   2.973n ±  0%  -20.99% (p=0.002 n=6)
Dot/10_dimensions-8      4.405n ± 0%   3.428n ±  2%  -22.19% (p=0.002 n=6)
Dot/12_dimensions-8      5.010n ± 6%   3.854n ±  0%  -23.08% (p=0.002 n=6)
Dot/16_dimensions-8      6.260n ± 0%   4.882n ±  1%  -22.01% (p=0.002 n=6)
Dot/24_dimensions-8      8.771n ± 5%   5.834n ±  3%  -33.49% (p=0.002 n=6)
Dot/30_dimensions-8     10.650n ± 5%   7.348n ±  2%  -31.01% (p=0.002 n=6)
Dot/32_dimensions-8     11.290n ± 4%   5.807n ±  0%  -48.56% (p=0.002 n=6)
Dot/128_dimensions-8     48.42n ± 0%   11.49n ± 14%  -76.27% (p=0.002 n=6)
Dot/256_dimensions-8     89.38n ± 0%   19.13n ±  0%  -78.60% (p=0.002 n=6)
Dot/300_dimensions-8    103.50n ± 0%   22.34n ±  3%  -78.41% (p=0.002 n=6)
Dot/384_dimensions-8    129.40n ± 2%   26.79n ±  0%  -79.30% (p=0.002 n=6)
Dot/600_dimensions-8    197.10n ± 0%   40.65n ±  2%  -79.38% (p=0.002 n=6)
Dot/768_dimensions-8    249.95n ± 6%   49.94n ±  0%  -80.02% (p=0.002 n=6)
Dot/1024_dimensions-8   330.05n ± 0%   65.10n ±  3%  -80.27% (p=0.002 n=6)
geomean                  19.37n        8.699n        -55.09%
```

**Review by trengrj — APPROVED** [review]

**Pull request merged**